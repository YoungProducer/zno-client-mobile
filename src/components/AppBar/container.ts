/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Declare main types and interfaces for AppBar component.
 * Create function which connect actions
 * and/or variables from the redux-store to the component.
 */

/** External imports */
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { green } from '@material-ui/core/colors';

/** Application's imports */

/** Define Material ui styles */
const styles = (theme: Theme) =>
    createStyles({
        root: {
            background: green[700],
            width: '100%',
            height: 50,
        },
        container: {
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            height: 'inherit',
        },
        navLink: {
            color: '#fff',
            fontSize: '1.2rem',
            marginRight: theme.spacing(3),
            '&:last-child': {
                marginRight: 0,
            },
        },
    });

/** Props which component get from the parent */
interface IOwnProps extends WithStyles<typeof styles>{}

/** Declare type which describe all props pushed to the component */
export type TAppBarProps =
    IOwnProps;

/** Export function to configure component via HOC */
export default compose(
    withStyles(styles, { withTheme: true }),
);
