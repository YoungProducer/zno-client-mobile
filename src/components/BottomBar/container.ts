/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Container for BottomBar component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { selectIsLoggedIn } from 'store/selectors/auth';
import { fetchLogoutAction } from 'store/actionCreators/auth';
import { RootState, setTestSuiteFinishedAction } from 'store/slices';

interface IStateProps {
    loggedIn: boolean;
}

interface IDispatchProps {
    logout: () => void;
    setTestSuiteFinished: (value: boolean) => void;
}

export type TBottomBarProps =
    IStateProps
    & IDispatchProps;

const mapStateToProps = (state: RootState): IStateProps => ({
    loggedIn: selectIsLoggedIn(state),
});

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    logout: () => dispatch(fetchLogoutAction()),
    setTestSuiteFinished: (finished: boolean) =>
        dispatch(setTestSuiteFinishedAction(finished)),
});

export default connect<IStateProps, IDispatchProps>(
    mapStateToProps,
    mapDispatchToProps,
);
