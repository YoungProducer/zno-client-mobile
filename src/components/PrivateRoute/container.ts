/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Define main types and interface related to PrivateRoute component.
 * Create function which connect variables from the redux store.
 */

/**External imports */
import { ReactNode } from 'react';
import { connect } from 'react-redux';
import { RouteProps } from 'react-router-dom';

// Applicaiton's imports
import { selectIsLoggedIn } from 'store/selectors/auth/signin';
import { RootState } from 'store/slices';

/**
 * Props which component get from the parent component.
 */
interface IOwnProps extends RouteProps {
    children: ReactNode;
}

/**
 * Props which component get from the redux store.
 */
interface IStateProps {
    isLoggedIn: boolean;
}

/**
 * Declare type which describe all props pushed to the component.
 */
export type TPrivateRoute =
    IOwnProps
    & IStateProps;

/**
 * Define function mapStateToProps.
 * Function select variables from the redux store.
 */
const mapStateToProps = (state: RootState): IStateProps => ({
    isLoggedIn: selectIsLoggedIn(state),
});

/**
 * Export function which connect actions
 * or/and variables from the redux store to component.
 */
export default connect<IStateProps, {}, IOwnProps>(
    mapStateToProps,
);
