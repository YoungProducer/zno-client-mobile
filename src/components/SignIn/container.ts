/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 29 February 2020
 *
 * Define main interfaces and types for SignIn component.
 * Created function which connect actions
 * and/or variables from the redux-store to the component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { fetchSignInAction } from 'store/actionCreators/auth';
import {
    setSignInErrorFieldsToDefaultAction,
    setSignInFieldsMessagesToDefaultAction,
    ISignInErrorFields,
    ISignInFieldsMessages,
} from 'store/slices/auth';
import {
    selectSignInLoading,
    selectSignInErrorFields,
    selectSignInFieldsMessages,
} from 'store/selectors/auth';
import { selectIsLoggedIn } from 'store/selectors/auth';
import { IFetchSignInActionCredentials } from 'utils/verify-credentials';
import { RootState } from 'store/slices';

/** Props which component get from the parent */
interface IOwnProps {}

/** Props which component get from the redux-store */
interface IStateProps {
    isLoggedIn: boolean;
    loading: boolean;
    errorFields: ISignInErrorFields;
    fieldsMessages: ISignInFieldsMessages;
}

/** Props(actions) which component can dispatch */
interface IDispatchProps {
    fetchSignIn: (credentials: IFetchSignInActionCredentials) => void;
    setSignInErrorFieldsToDefault: () => void;
    setSignInFieldsMessagesToDefault: () => void;
}

/** Declare type which describe all props pushed to the component */
export type TSignInProps =
    IOwnProps
    & IStateProps
    & IDispatchProps;

/** Function which connect variables from the redux-store to component */
const mapStateToProps = (state: RootState): IStateProps => ({
    isLoggedIn: selectIsLoggedIn(state),
    loading: selectSignInLoading(state),
    errorFields: selectSignInErrorFields(state),
    fieldsMessages: selectSignInFieldsMessages(state),
});

/** Function which connect actions to the component */
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    fetchSignIn: (credentials: IFetchSignInActionCredentials) =>
        dispatch(fetchSignInAction(credentials)),

    setSignInErrorFieldsToDefault: () =>
        dispatch(setSignInErrorFieldsToDefaultAction()),

    setSignInFieldsMessagesToDefault: () =>
        dispatch(setSignInFieldsMessagesToDefaultAction()),
});

/**
 * Export function which connect actions
 * and/or variables from the redux-store to the component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateToProps,
    mapDispatchToProps,
);
