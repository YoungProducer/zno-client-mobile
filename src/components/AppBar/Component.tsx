/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Header component which contain main routes and another information or actions.
 */

/** External imports */
import React from 'react';
import CssBaseLine from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';
import AppBar from '@material-ui/core/AppBar';

/** Application's imports */
import { TAppBarProps } from './container';
import NavigationLink from 'components/NavigationLink';

/** Create component */
const Component = (props: TAppBarProps) => {
    const { classes } = props;

    return (
        <React.Fragment>
            <CssBaseLine />
            <AppBar
                position='static'
                className={classes.root}
            >
                <Container
                    maxWidth='lg'
                    className={classes.container}
                >
                    <NavigationLink
                        navLink={{
                            to: '/',
                        }}
                        className={classes.navLink}
                        data-testid='home-navlink'
                    >
                        Головна
                    </NavigationLink>
                    <NavigationLink
                        navLink={{
                            to: '/subject-selection',
                        }}
                        className={classes.navLink}
                        data-testid='tests-navlink'
                    >
                        Тести
                    </NavigationLink>
                </Container>
            </AppBar>
        </React.Fragment>
    );
};

export default Component;
