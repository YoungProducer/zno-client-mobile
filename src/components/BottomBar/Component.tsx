/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Component which display main navigation of app.
 */

/** External imports */
import React, { useState, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import classNames from 'classnames';
import Collapse from '@material-ui/core/Collapse';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import ExpandLessIcon from '@material-ui/icons/ExpandLess';
import HomeIcon from '@material-ui/icons/Home';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import PersonIcon from '@material-ui/icons/Person';
import { makeStyles, createStyles, Theme } from '@material-ui/core';

/** Application's imports */
import { TBottomBarProps } from './container';

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: 'absolute',
            minHeight: 70,
            borderRadius: 9,
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))',
            bottom: theme.spacing(1),
            left: theme.spacing(1),
            right: theme.spacing(1),
            transform: `translateY(0)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        rootHidden: {
            transform: `translateY(60%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        innerContainer: {
            width: '100%',
            height: '100%',
        },
        actionsContainer: {
            width: '100%',
            height: '100%',
            minHeight: 70,
            display: 'flex',
            justifyContent: 'space-around',
            alignItems: 'center',
        },
        expandIconButton: {
            zIndex: 10,
            position: 'absolute',
            left: '50%',
            transform: `translate(-50%, -60%)`,
            borderRadius: '50%',
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.05))',
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
            '&:hover': {
                background: '#fff',
            },
        },
        expandIconButtonHidden: {
            transform: `translate(-50%, -35%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIcon: {
            transform: `rotateX(180deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIconHidden: {
            transform: `rotateX(0deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButton: {
            background: '#fff',
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 300,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButtonActive: {
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0.16)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 500,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIcon: {
            color: '#69deac',
            width: 35,
            height: 35,
        },
        expandActions: {
            padding: theme.spacing(1),
            display: 'flex',
            justifyContent: 'space-between',
        },
        button: {
            color: '#fff',
            background: '#656cff',
            borderRadius: 4,
            width: '45%',
        },
    }));

const Component = ({ loggedIn, logout }: TBottomBarProps) => {
    const classes = useStyles({});

    const history = useHistory();
    const location = useLocation();

    const [hidden, setHidden] = useState<boolean>(false);
    const [expand, setExpand] = useState<boolean>(false);

    const toggleHidden = () => {
        if (expand) {
            setExpand(false);
            window.setTimeout(() => {
                setHidden(!hidden);
            }, 300);
        } else {
            setHidden(!hidden);
        }
    };
    const toggleExpand = () => setExpand(!expand);

    const handleSignIn = () => history.push('/auth/signin');
    const handleSignUp = () => history.push('/auth/signup');
    const handleHome = () => history.push('/');
    const handleProfile = () => history.push('/user/profile');

    const activeIcons = useMemo(() => {
        const { pathname } = location;

        return {
            home: pathname === '/' || pathname === '/subject-selection',
            profile: pathname === '/user/profile',
        };
    }, [location]);

    return (
        <div
            className={classNames(classes.root, {
                [classes.rootHidden]: hidden,
            })}
        >
            <IconButton
                className={classNames(classes.expandIconButton, {
                    [classes.expandIconButtonHidden]: hidden,
                })}
                onClick={toggleHidden}
            >
                <ExpandLessIcon className={classNames(classes.expandIcon, {
                    [classes.expandIconHidden]: hidden,
                })}/>
            </IconButton>
            <div className={classes.innerContainer}>
                <div className={classes.actionsContainer}>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.profile,
                        })}
                        onClick={handleProfile}
                    >
                        <PersonIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.home,
                        })}
                        onClick={handleHome}
                    >
                        <HomeIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: expand,
                        })}
                        onClick={toggleExpand}
                    >
                        <MoreHorizIcon className={classes.actionIcon}/>
                    </IconButton>
                </div>
                <Collapse in={expand} unmountOnExit>
                    <div className={classes.expandActions}>
                        { loggedIn && (
                            <Button
                                className={classes.button}
                                variant='contained'
                                disableElevation
                                onClick={logout}
                            >
                                Вийти
                            </Button>
                        )}
                        { !loggedIn && (
                            <>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignIn}
                                >
                                    Увійти
                                </Button>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignUp}
                                >
                                    Зареєструватися
                                </Button>
                            </>
                        )}
                    </div>
                </Collapse>
            </div>
        </div>
    );
};

export default Component;
