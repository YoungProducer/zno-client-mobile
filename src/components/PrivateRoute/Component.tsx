/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Permission route which automaticaly redirect from protected pages if user is not logged in to SignIn page.
 */

/** External imports */
import React from 'react';
import { Redirect, Route } from 'react-router-dom';

/** Application's imports */
import { TPrivateRoute } from './container';

/** Create component */
const Component = ({ isLoggedIn, children, ...rest }: TPrivateRoute) => (
    <Route
        {...rest}
        render={({ location }) => {
            const pathname = location
                ? `/auth/signin?next=${location.pathname}${location.search}`
                : '/auth/signin';

            return isLoggedIn
                ? children
                : <Redirect
                    to={{
                        pathname: '/auth/signin',
                        state: { from: location },
                    }}
                />;
        }
        }
    />
);

export default Component;
