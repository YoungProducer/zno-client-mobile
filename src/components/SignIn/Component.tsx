/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 February 2020
 *
 * Component which allows user to make sign in
 */

/** External imports */
import React, { useCallback, useState, useEffect } from 'react';
import classNames from 'classnames';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import EmailIcon from '@material-ui/icons/Email';
import LockIcon from '@material-ui/icons/Lock';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { useHistory } from 'react-router-dom';

/** Application's imports */
import { TSignInProps } from './container';
import NavigationLink from 'components/NavigationLink';
import Input from 'components/custom/Input';
import Checkbox from 'components/custom/Checkbox';
import Logo from 'img/logo';

/** Define classes as hook */
const useStyles = makeStyles((theme: Theme) => createStyles({
    container: {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        width: '100%',
        boxSizing: 'border-box',
        padding: theme.spacing(1),
    },
    root: {
        borderRadius: 4,
        width: '100%',
        filter: `drop-shadow(2.828px 2.828px 3.5px rgba(227,207,207,0.9))`,
        backgroundColor: '#ffffff',
        textAlign: 'center',
    },
    innerContainer: {
        borderRadius: 4,
        backgroundColor: '#f4f5f7',
        // padding: theme.spacing(4),
    },
    logoContainer: {
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    textField: {
        width: 230,
    },
    fields: {
        paddingBottom: theme.spacing(3),
    },
    actions: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    block: {
        width: '100%',
        margin: 0,
        '& .MuiGrid-item': {
            width: '100%',
        },
    },
    buttonBlock: {
        width: '100%',
        marginTop: theme.spacing(2),
    },
    button: {
        filter: `drop-shadow(0px 3px 2.5px rgba(0,0,0,0.16))`,
    },
    link: {
        color: theme.palette.primary.main,
    },
    icon: {
        color: '#555',
    },
    inputRoot: {
        width: '100%',
        height: 60,
        background: `#fff`,
        filter: `drop-shadow(0px 3px 2.5px rgba(0,0,0,0.16))`,
        borderRadius: 4,
        paddingLeft: theme.spacing(2),
        paddingRight: theme.spacing(2),
    },
    checkBox: {
        width: '100%',
    },
}));

/**
 * Create hook to make component stateless.
 * Compute all values from text fields.
 * Returns fields data and button onClick property.
 */
const useSignInElements = (props: TSignInProps) => {
    const {
        isLoggedIn,
        errorFields,
        fieldsMessages,
        fetchSignIn,
        setSignInErrorFieldsToDefault,
        setSignInFieldsMessagesToDefault,
    } = props;

    const history = useHistory();

    useEffect(() => {
        if (isLoggedIn) {
            const next = history.location.state
                ? `${(history.location.state as any).from.pathname}${(history.location.state as any).from.search}`
                : undefined;

            history.push(next || '/');
        }
    }, [isLoggedIn]);

    const clearFields = () => {
        setSignInErrorFieldsToDefault();
        setSignInFieldsMessagesToDefault();
    };

    /** User email */
    const [email, setEmail] = useState<string>('');
    const handleChangeEmail =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setEmail(event.target.value);
            if (errorFields.email) {
                clearFields();
            }
        }, [errorFields.email]);

    /** User password */
    const [password, setPassword] = useState<string>('');
    const handleChangePassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setPassword(event.target.value);
            if (errorFields.password) {
                clearFields();
            }
        }, [errorFields.password]);

    /** Remember checkbox */
    const [remember, setRemember] = useState<boolean>(false);
    const handleChangeRemember =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setRemember(event.target.checked);
        }, []);

    /** SignUp button handle */
    const handleFetchSignIn = useCallback(() => fetchSignIn({
        email,
        password,
        remember,
    }), [email, password, remember]);

    return {
        emailField: {
            onChange: handleChangeEmail,
            value: email,
            error: errorFields.email,
            helperText: fieldsMessages.email,
        },
        passwordField: {
            onChange: handleChangePassword,
            value: password,
            error: errorFields.password,
            helperText: fieldsMessages.password,
        },
        rememberCheckbox: {
            onChange: handleChangeRemember,
            value: remember,
        },
        signInButton: {
            onClick: handleFetchSignIn,
        },
    };
};

/** Create component */
const Component = (props: TSignInProps) => {
    /** Create styles */
    const classes = useStyles({});

    /** Get fields data from hook */
    const { emailField, passwordField, rememberCheckbox, signInButton } = useSignInElements(props);

    return (
        <div
            className={classNames('auth-background', classes.container)}
        >
            <Paper
                className={classes.root}
                elevation={0}
                square
            >
                <div className={classes.logoContainer}>
                    <Logo />
                </div>
                <div className={classes.innerContainer}>
                    <Grid
                        container
                        direction='column'
                        alignItems='center'
                        justify='flex-start'
                        item
                        spacing={3}
                        className={classes.block}
                    >
                        <Grid item>
                            <Input
                                color='primary'
                                type='email'
                                placeholder='Емеїл'
                                startAdornment={<EmailIcon className={classes.icon}/>}
                                data-testid='signin-email-input'
                                rootClassName={classes.inputRoot}
                                {...emailField}
                            />
                        </Grid>
                        <Grid item>
                            <Input
                                color='primary'
                                type='password'
                                placeholder='Пароль'
                                startAdornment={<LockIcon className={classes.icon}/>}
                                data-testid='signin-password-input'
                                rootClassName={classes.inputRoot}
                                {...passwordField}
                            />
                        </Grid>
                        <Grid item style={{
                            width: 'inherit',
                        }}>
                            <Checkbox
                                {...rememberCheckbox}
                                label={`Запам'ятати мене`}
                                formControl={{
                                    className: classes.checkBox,
                                }}
                            />
                        </Grid>
                        <Grid item>
                            <Button
                                color='primary'
                                variant='contained'
                                disableElevation
                                className={classes.button}
                                data-testid='signin-button'
                                { ...signInButton }
                            >
                                Увійти
                            </Button>
                        </Grid>
                        <Grid item>
                            <Typography align='center'>
                                Не зареєстровані?
                                <NavigationLink
                                    navLink={{
                                        to: '/auth/signup',
                                    }}
                                    className={classes.link}
                                >
                                    Зареєструватися
                                </NavigationLink>
                            </Typography>
                        </Grid>
                    </Grid>
                </div>
            </Paper>
        </div>
    );
};

export default Component;
