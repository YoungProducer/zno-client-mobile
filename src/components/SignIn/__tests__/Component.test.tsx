/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 29 February 2020
 *
 * Create test suites for SignIn component.
 */

/** External imports */
import React from 'react';
import { shallow, mount } from 'enzyme';

/** Application's imports */
import Component from '../Component';
import { TSignInProps } from '../container';

/** Create mock function for useHistory hook */
const push = jest.fn();

/** Mock react-router-dom module */
jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(() => ({
        push,
    })),
}));

describe('SignIn component', () => {
    /** Mock function for props */
    const setSignInFieldsMessagesToDefault = jest.fn();
    const fetchSignIn = jest.fn();
    const setSignInErrorFieldsToDefault = jest.fn();

    /** Define required props */
    const requiredProps: TSignInProps = {
        fetchSignIn,
        setSignInErrorFieldsToDefault,
        setSignInFieldsMessagesToDefault,
        loading: false,
        errorFields: {
            email: false,
            password: false,
        },
        fieldsMessages: {
            email: 'foo',
            password: 'bar',
        },
        isLoggedIn: false,
    };

    beforeEach(() => {
        (requiredProps.fetchSignIn as jest.Mock).mockReset();
        (requiredProps.setSignInErrorFieldsToDefault as jest.Mock).mockReset();
        (requiredProps.setSignInFieldsMessagesToDefault as jest.Mock).mockReset();
    });

    test('Is match snapshot', () => {
        /** Render component in shallow */
        const root = shallow(<Component {...requiredProps}/>);

        /** Assert component matches snapsot */
        expect(root).toMatchSnapshot();
    });

    test('Is fetchSignIn will be called with right args', () => {
        /** Create dom tree */
        const tree = shallow(<Component {...requiredProps}/>);

        /** Simulate onChange events*/
        tree.find(`[data-testid='signin-email-input']`).simulate('change', {
            target: {
                value: 'bar',
            },
        });

        tree.find(`[data-testid='signin-password-input']`).simulate('change', {
            target: {
                value: 'foo',
            },
        });

        /** Simulate signin button click */
        tree.find(`[data-testid='signin-button']`).simulate('click');

        /** Assert fetchSignIn be called with right values */
        expect(fetchSignIn).toBeCalledWith({
            email: 'bar',
            password: 'foo',
            remember: false,
        });
    });

    test('Is clear function will be called when fields have errors after email onChange event', () => {
        /** Create dom tree */
        const tree = shallow(<Component
            {...requiredProps}
            errorFields={{
                email: true,
                password: true,
            }}
            fieldsMessages={{
                email: 'foo',
                password: 'bar',
            }}
        />);

        /** Simulate onChange event */
        tree.find(`[data-testid='signin-email-input']`).simulate('change', {
            target: {
                value: 'bar',
            },
        });

        /** Assert setSignInErrorFieldsToDefault have been called and just one time*/
        expect(setSignInErrorFieldsToDefault).toHaveBeenCalled();
        expect(setSignInErrorFieldsToDefault).toHaveBeenCalledTimes(1);

        /** Assert setSignInFieldsMessagesToDefault have been called and just one time */
        expect(setSignInFieldsMessagesToDefault).toHaveBeenCalled();
        expect(setSignInFieldsMessagesToDefault).toHaveBeenCalledTimes(1);
    });

    test('Is clear function will be called when fields have errors after password onChange event', () => {
        /** Create dom tree */
        const tree = shallow(<Component
            {...requiredProps}
            errorFields={{
                email: true,
                password: true,
            }}
            fieldsMessages={{
                email: 'foo',
                password: 'bar',
            }}
        />);

        /** Simulate onChange event */
        tree.find(`[data-testid='signin-password-input']`).simulate('change', {
            target: {
                value: 'bar',
            },
        });

        /** Assert setSignInErrorFieldsToDefault have been called and just one time*/
        expect(setSignInErrorFieldsToDefault).toHaveBeenCalled();
        expect(setSignInErrorFieldsToDefault).toHaveBeenCalledTimes(1);

        /** Assert setSignInFieldsMessagesToDefault have been called and just one time */
        expect(setSignInFieldsMessagesToDefault).toHaveBeenCalled();
        expect(setSignInFieldsMessagesToDefault).toHaveBeenCalledTimes(1);
    });

    test('Is text fields property error equals true if fields filled incorrectly', () => {
        /** Create dom tree */
        const tree = shallow(<Component
            {...requiredProps}
            errorFields={{
                email: true,
                password: true,
            }}
            fieldsMessages={{
                email: 'foo',
                password: 'bar',
            }}
            setSignInErrorFieldsToDefault={setSignInErrorFieldsToDefault}
        />);

        /** Assert that both text fields have errors */
        expect((tree.find(`[data-testid='signin-email-input']`).props() as any).error).toBeTruthy();
        expect((tree.find(`[data-testid='signin-password-input']`).props() as any).error).toBeTruthy();

        /** Simulate onChange event */
        tree.find(`[data-testid='signin-email-input']`).simulate('change', { target: { value: 'foo' } });

        /** Assert functions for setting errors to default have been called */
        expect(setSignInErrorFieldsToDefault).toHaveBeenCalled();
        expect(setSignInFieldsMessagesToDefault).toHaveBeenCalled();
    });
});
