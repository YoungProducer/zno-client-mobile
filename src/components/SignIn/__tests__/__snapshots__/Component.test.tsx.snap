// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component Is match snapshot 1`] = `
<div
  className="auth-background"
>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-container-2"
    maxWidth="lg"
  >
    <WithStyles(ForwardRef(Paper))
      className="makeStyles-root-3"
      elevation={0}
      square={true}
    >
      <div
        className="makeStyles-logoContainer-5"
      >
        <Component />
      </div>
      <div
        className="makeStyles-innerContainer-4"
      >
        <WithStyles(ForwardRef(Grid))
          alignItems="center"
          className="makeStyles-block-9"
          container={true}
          direction="column"
          item={true}
          justify="flex-start"
          spacing={3}
        >
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <Component
              color="primary"
              data-testid="signin-email-input"
              error={false}
              helperText="foo"
              onChange={[Function]}
              placeholder="Емеїл"
              startAdornment={
                <Memo(EmailIcon)
                  className="makeStyles-icon-13"
                />
              }
              type="email"
              value=""
            />
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <Component
              color="primary"
              data-testid="signin-password-input"
              error={false}
              helperText="bar"
              onChange={[Function]}
              placeholder="Пароль"
              startAdornment={
                <Memo(LockIcon)
                  className="makeStyles-icon-13"
                />
              }
              type="password"
              value=""
            />
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <Component
              label="Запам'ятати мене"
              onChange={[Function]}
              value={false}
            />
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <WithStyles(ForwardRef(Button))
              className="makeStyles-button-11"
              color="primary"
              data-testid="signin-button"
              disableElevation={true}
              onClick={[Function]}
              variant="contained"
            >
              Увійти
            </WithStyles(ForwardRef(Button))>
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <WithStyles(ForwardRef(Typography))
              align="center"
            >
              Не зареєстровані?
              <NavigationLink
                className="makeStyles-link-12"
                navLink={
                  Object {
                    "to": "/auth/signup",
                  }
                }
              >
                Зареєструватися
              </NavigationLink>
            </WithStyles(ForwardRef(Typography))>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </div>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Container))>
</div>
`;
